### DISCUSSION TITLE
How “Draw” is working in Server side angular Datatables

### DISCUSSION THREAD

**Discussion by RajasekharKp** [l-lin/angular-datatables#1555] [Q&A]
Trying to implement, server side angular datatables for implementing server side pagination. Data is populated from API, but datatable properties are not working(sort, search, UI pagination). I observed in the example API, `draw `keyword is dynamically updating for every pagination click. How can we send draw parameter from UI ? what is the importance of draw key in the response. is `draw `key is mandatory in JSON response ? could you please help me on this issue,

I took below example for reference, but not sure how draw is updating for every click. Please suggest me.

[https://stackblitz.com/edit/angular-datatable-server-side-dwxpsz](url)
[http://l-lin.github.io/angular-datatables/#/basic/server-side-angular-way](url)

**Comment by shanmukhateja** [comment]
Hi,

If you take a look at [Server side processing](https://datatables.net/manual/server-side) in datatables.net documentation, you'll find that `draw` counter is a mandatory parameter required by datatables.net. It is simply a counter that keeps track of each Ajax request to ensure the data you see matches the input you gave. For example, say you have rendered your data to the table via Ajax, you'll notice `draw` value will be set to `1` (as in first Ajax request). Now, navigate to the next page using pagination and you'll notice `draw` becomes `2`. It'll prevent edge-cases where say you try to render the 1st page of your table by recursively clicking on "Prev" button but ending up with the results of 2nd page as these requests _can run out of sequence_ acc. to documentation.

I believe you're looking at this the wrong way. The reason why pagination isn't working is because you're sending incorrect values to `recordsTotal` or `recordsFiltered` or both. Sort and search need to be implemented on your API server by filtering your database using `dataTablesParameters` on the POST request. 

Check your API server.
